### PR TITLE
Directory support and separate file descriptor spaces in the execution engine.

### PR DESCRIPTION
**Pull request by ShaleXIONG**
Add minimum directory support in the execution engine and separate file descriptor space as per client session and program execution.
- File descriptors are now local to client (TLS) sessions and program executions. Directories will be pre-opened with appropriate capabilities as specified in the policy and passed to clients and programs.
- File System is now organized as a shareable and mutable inode table, which includes all data, and several file system handlers for accessing the inode table.
- The filesystem now contains a directory tree structure, where each `Inode` can be a file or a directory. Remove the old `path_table`.
- Implement `path_create_directory` and rework on `path_open`.
- Rework on Freestand-execution-engine. `enable_clock`, `dump_stdout`, `dump_stderr` now are flags, rather than parameters, the default is false. `input` parameter now accept a directory, which will load recursively ALL the files from the host directory to the VFS root. `output` does the reverse. Both `input` and `output` also grant read and write capabilities to the program. `program` parameter now does NOT copy the program from the host to the VFS (please use `input` to do it, as a program is just a file), but only tell the free-EE where is the program in VFS.
- Improve code quality. Move several functionalities from the file system top level to the inode level.

Minor
- Improve the freestanding execution engine.
- Improve several test cases. The idash2017 test now directly read all files in a directory and apply computation on all the files; similar to macd and privacity-inter-set-sum. Parameters in the test_templates in both veracruz-server-test and veracruz-test now include the remote paths, which indicates where the provisioned program or data should be store. 
- Update the policy format. Add `hashes` for specfying files, inc. binaries, and their hashes. The runtime ensures the file matches the hash, if it is present in the policy.

### CODE REVIEW

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 59630d6909480ad678485594055816f9071591fa
* Result: FAILED
* [Build Logs](https://01evtei465.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=c27d83dd-5085-407c-9761-492734c41e9c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: e7b2e360f69ecc41553e1817d6ebd61d0a5434ac
* Result: FAILED
* [Build Logs](https://01evtei465.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=985d7045-8ca6-42b8-9ea0-1d50a4637a53%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 3a47000024114df41b4e001b7babc0af87a71c6d
* Result: FAILED
* [Build Logs](https://01evtei465.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=4b942e98-0654-441b-88b2-5d13fd9e5970%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 952b4d4ca7f1695af1c5adf8d6b00e4827afc25f
* Result: FAILED
* [Build Logs](https://01evtei465.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=b8ed004d-b170-492e-a270-bb2b37e59cfd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: f10c968d4cc8d2b1de9d559ed2089d40a3753f89
* Result: FAILED
* [Build Logs](https://01evtei465.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=5c232b88-c8a3-4768-bb73-b51bd154ee77%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 7b9cda904b4c2540c35ecc0c986b5cd4b86ac0df
* Result: FAILED
* [Build Logs](https://01evtei465.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=8cbc3b28-afba-4ce4-870a-d739577b5c2e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: e14f6f105ce56e763d1ef1bcfa759fd79381dfbb
* Result: FAILED
* [Build Logs](https://01evtei465.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=21dde477-03cd-40b2-81ce-33246768439f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by veracruz-project-owner** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Veracruz
* Commit ID: 62bd22175b3b2f678aca1b279777130a50061907
* Result: FAILED
* [Build Logs](https://01evtei465.execute-api.eu-west-2.amazonaws.com/Prod/buildlogs?key=d9d14203-b13b-4aa7-965b-9d07c0d1766a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: execution-engine/Cargo.toml
Comment (dominic-mulligan-arm) [inline execution-engine/Cargo.toml:36]: We should commit to a particular version of `log` here.

File: execution-engine/src/fs.rs
Comment (dominic-mulligan-arm) [inline execution-engine/src/fs.rs:28]: I think we should avoid these types of wildcard import, in favour of being explicit about what it is we are importing from a particular library.

File: execution-engine/src/fs.rs
Comment (dominic-mulligan-arm) [inline execution-engine/src/fs.rs:123]: This can be marked as `inline`.

File: execution-engine/src/fs.rs
Comment (dominic-mulligan-arm) [inline execution-engine/src/fs.rs:117]: This can be marked as `inline`.

File: execution-engine/src/fs.rs
Comment (dominic-mulligan-arm) [inline execution-engine/src/fs.rs:101]: This can be marked as `inline`.

File: execution-engine/src/fs.rs
Comment (dominic-mulligan-arm) [inline execution-engine/src/fs.rs:95]: This can be marked as `inline`.

File: execution-engine/src/fs.rs
Comment (dominic-mulligan-arm) [inline execution-engine/src/fs.rs:72]: This can be marked as `inline`.

File: execution-engine/src/fs.rs
Comment (dominic-mulligan-arm) [inline execution-engine/src/fs.rs:664]: This can be marked as `inline`.

File: execution-engine/src/fs.rs
Comment (dominic-mulligan-arm) [inline execution-engine/src/fs.rs:1034]: Are these logging messages just development aids, or will they stay in the code?

File: execution-engine/src/fs.rs
Comment (dominic-mulligan-arm) [inline execution-engine/src/fs.rs:1502]: These fixed paths for the parent and current directories can probably be made constants.

File: execution-engine/src/fs.rs
Comment (ShaleXIONG) [reply execution-engine/src/fs.rs:1034]: It will be erased before merging.

File: execution-engine/Cargo.toml
Comment (ShaleXIONG) [reply execution-engine/Cargo.toml:36]: It will be removed before merging, this is for debugging during dev.

File: policy-utils/src/policy.rs
Comment (dominic-mulligan-arm) [inline policy-utils/src/policy.rs:100]: Rustdoc comment has been demoted to a normal Rust comment, here.

File: policy-utils/src/policy.rs
Comment (dominic-mulligan-arm) [inline policy-utils/src/policy.rs:147]: Can this be deleted?

File: policy-utils/src/policy.rs
Comment (dominic-mulligan-arm) [inline policy-utils/src/policy.rs:280]: Can this be deleted?

File: policy-utils/src/policy.rs
Comment (dominic-mulligan-arm) [inline policy-utils/src/policy.rs:389]: Can this be deleted?

File: sdk/data-generators/private-set-intersection-sum-generator/src/main.rs
Comment (dominic-mulligan-arm) [inline sdk/data-generators/private-set-intersection-sum-generator/src/main.rs:104]: Can we have some more elegant error handling here, rather than `unwrap()` failing?  e.g. use `map_err()` and `eprintln!()` to explain what went wrong and kill the program with an appropriate error code.

File: veracruz-client/src/veracruz_client.rs
Comment (dominic-mulligan-arm) [inline veracruz-client/src/veracruz_client.rs:264]: Largely the same comments here as above.

File: veracruz-client/src/veracruz_client.rs
Comment (dominic-mulligan-arm) [inline veracruz-client/src/veracruz_client.rs:282]: Similar comments here regarding the type of `path` as above, and a lack of `rustdoc` comments on this public function.

File: veracruz-client/src/veracruz_client.rs
Comment (dominic-mulligan-arm) [inline veracruz-client/src/veracruz_client.rs:303]: Similar comments here regarding the type of `path` as above, and a lack of `rustdoc` comments on this public function.

**Review by dominic-mulligan-arm — CHANGES_REQUESTED** [review]
Good work, @ShaleXIONG!  Just a few minor comments.  Also, have you made sure that this compiles for `nitro` in lieu of us having an explicit CI target for that?

File: execution-engine/src/wasi/common.rs
Comment (egrimley-arm) [inline execution-engine/src/wasi/common.rs:345]: Some words missing in this sentence!

File: sdk/freestanding-execution-engine/src/main.rs
Comment (egrimley-arm) [inline sdk/freestanding-execution-engine/src/main.rs:239]: Is this equivalent to `let dump_stdout = DEFAULT_DUMP_STDOUT || matches.is_present("dump-stdout");`? (Similarly for `dump_stderr`.) And I'm wondering if there's any point in having `DEFAULT_DUMP_STDOUT` in the code at all seeing as it will probably remain `false` for ever.

File: sdk/freestanding-execution-engine/src/main.rs
Comment (egrimley-arm) [inline sdk/freestanding-execution-engine/src/main.rs:387]: I think you might need `io::stdout()::flush()` or something like that to avoid the two getting mixed up if `stdout` and `stderr` are going to the same place which isn't a terminal.

Whenever you have finished writing to `stdout` and are about to write to `stderr` (and perhaps the other way round if you are feeling nervous) would probably be a good time to `flush`.

File: execution-engine/src/lib.rs
Comment (egrimley-arm) [inline execution-engine/src/lib.rs:67]: executuon -> execution

**Review by dominic-mulligan-arm — DISMISSED** [review]

**Review by dominic-mulligan-arm — APPROVED** [review]

**Comment by dominic-mulligan-arm** [comment]
+1 reached (and @egrimley-arm also happy to merge, after a personal conversation).  Merging.

**Pull request merged**
